domain/src/test/java/org/oppia/domain/audio/AudioPlayerControllerTest.kt

```diff
@@ -0,0 +1,470 @@
+  }
+
+  @Test
+  fun testObserver_observeWithFragment_invokeRelease_checkRemoveObservers() {
+    val playProgress = audioPlayerController.initializeMediaPlayer(TEST_URL)
+    playProgress.observe(fragment, mockAudioPlayerObserver)
+
+    audioPlayerController.play()
+    audioPlayerController.releaseMediaPlayer()
+
+    assertThat(shadowMediaPlayer.state).isEqualTo(ShadowMediaPlayer.State.END)
+    assertThat(playProgress.hasObservers()).isFalse()
+  }
+
+  /** Ensure test dispatcher is running. If test does not hang, behavior is correct */
+  @Test
+  @ExperimentalCoroutinesApi
+  fun testScheduling_preparePlayer_invokePauseAndAdvance_checkNoNewUpdates()
+      = runBlockingTest(coroutineContext) {
+    arrangeMediaPlayer()
+
+    audioPlayerController.play()
+    advanceTimeBy(2000)
+    audioPlayerController.pause()
+    advanceTimeBy(2000)
+
+    verify(mockAudioPlayerObserver, atLeastOnce()).onChanged(audioPlayerResultCaptor.capture())
+    assertThat(audioPlayerResultCaptor.value.getOrThrow().type).isEqualTo(PlayStatus.PAUSED)
+  }
+
+  /** Ensure test dispatcher is running. If test does not hang, behavior is correct */
+  @Test
+  @ExperimentalCoroutinesApi
+  fun testScheduling_preparePlayer_invokeCompletionAndAdvance_checkJobIsNull()
+      = runBlockingTest(coroutineContext) {
+    arrangeMediaPlayer()
+
+    audioPlayerController.play()
+    advanceTimeBy(2000)
+    shadowMediaPlayer.invokeCompletionListener()
+    advanceTimeBy(2000)
+
+    verify(mockAudioPlayerObserver, atLeastOnce()).onChanged(audioPlayerResultCaptor.capture())
+    assertThat(audioPlayerResultCaptor.value.getOrThrow().type).isEqualTo(PlayStatus.COMPLETED)
+  }
+
+  /** Ensure test dispatcher is running. If test does not hang, behavior is correct */
+  @Test
+  @ExperimentalCoroutinesApi
+  fun testScheduling_observeData_removeObserver_checkJobIsNull()
+      = runBlockingTest(coroutineContext) {
+    val playProgress = audioPlayerController.initializeMediaPlayer(TEST_URL)
+    playProgress.observeForever(mockAudioPlayerObserver)
+
+    audioPlayerController.play()
+    advanceTimeBy(2000)
+    playProgress.removeObserver(mockAudioPlayerObserver)
+    advanceTimeBy(2000)
+  }
+
+  /** Ensure test dispatcher is running. If test does not hang, behavior is correct */
+  @Test
+  @ExperimentalCoroutinesApi
+  fun testScheduling_addAndRemoveObservers_checkJobStates() = runBlockingTest(coroutineContext) {
+    val playProgress = audioPlayerController.initializeMediaPlayer(TEST_URL)
+
+    audioPlayerController.play()
+    advanceTimeBy(2000)
+
+    playProgress.observeForever(mockAudioPlayerObserver)
+    audioPlayerController.pause()
+
+    playProgress.removeObserver(mockAudioPlayerObserver)
+    audioPlayerController.play()
+    advanceTimeBy(2000)
+  }
+
+  @Test
+  fun testController_alreadyInitialized_initializePlayer_fails() {
+    arrangeMediaPlayer()
+
+    assertThrows(IllegalStateException::class) {
```

Comment:
Suggest verifying that correct exception is thrown by checking its message. E.g.:

```kotlin
val exception = assertThrows(IllegalStateException.class) { ... }
assertThat(exception).hasMessageThat().contains("Identifying details about reason")
```

Since IllegalStateException is generic, it's conceivable that the wrong one is actually being thrown here. Verifying the correct exception is thrown helps tighten up the behavior the test itself is verifying.